accepted/2022/FunctionPointers.md

```diff
@@ -0,0 +1,671 @@
+# Function Pointers
+#### Steve Harter
+#### December 20, 2022
+
+# Background
+For additional context, see:
+- [Original feature ask](https://github.com/dotnet/runtime/issues/11354)
+- [MetadataLoadContext feature ask](https://github.com/dotnet/runtime/issues/43791)
+- [7.0 introspection attempt](https://github.com/dotnet/runtime/pull/71516)
+- [C# function pointer spec](https://learn.microsoft.com/dotnet/csharp/language-reference/proposals/csharp-9.0/function-pointers)
+
+Function pointers, which have been in ECMA-335 and used by the CLR since the beginning were only recently supported by C# in C# v9 and .NET v6. However, the .NET type system has not been updated to directly support them and they are currently exposed as the `IntPtr` type:
+```cs
+Type type = typeof(delegate*<int, bool>); // System.IntPtr (currently)
+```
+
+Reflection also returns `IntPtr` for both runtime-reflection through `FieldInfo\ParameterInfo\PropertyInfo` and static-reflection through `MetadataLoadContext`:
+```cs
+  FieldInfo fi = typeof(MyClass).GetField("_fn");
+  Type type = fi.FieldType; // System.IntPtr (currently)
+
+  unsafe class MyClass
+  {
+      public delegate*<int, bool> _fn; // A field as a function pointer
+  }
+```
+
+Treating function pointers as `IntPtr` essentially makes reflection introspection impossible, blocking scenarios that care about function pointers. The proposal is to change from `IntPtr` to `Type` which is a **breaking change**:
+```cs
+Type type = typeof(delegate*<int, bool>); // System.Type (proposed design)
+bool isFunctionPointer = type.IsFunctionPointer; // A new method which returns 'true' here
+```
+
+Note that Mono returns "System.MonoFNPtrFakeClass" for the type name, not "IntPtr". This design applies to Mono as well and is expected that both runtimes are consistent going forward.
```

Comment:
@lambdageek Any concerns with the above? Looking across github I see relatively few mentions of `MonoFNPtrFakeClass` so assume this is likely to be uninteresting.

Reply:
I _think_ `MonoFNPtrFakeClass` is just used to fill in `MonoClass:name` (because we often try to use that field in debug output and the like and null-checking it everywhere is probably a pain).  That it happens to also be used when you call `type.FullName` or `type.ToString()` is incidental and we could change it.  Internally in mono function pointer types are quite distinct from other types - having new behavior for them shouldn't be an architectural problem (implementation might be another matter: there have been - and probably still are - many `switch` statements in the C code that don't have a good fn ptr branch)

@AaronRobinsonMSFT  this intro section is confusing given what comes later in the design.
You're not proposing `typeof(delegate*<int,bool>) == typeof(System.Type)` . The behavior today is that `typeof(delegate*<int,bool) == typeof(System.IntPtr)`, and what you want is an actual new instance of `System.Type` for each function pointer type, right?

Reply:
>  and what you want is an actual new instance of System.Type for each function pointer type, right?

Yes a separate instance for each unique function pointer (ignoring custom modifiers). I updated this intro section. Thanks